sympy/solvers/simplex.py

```diff
@@ -0,0 +1,966 @@
+    constraint that ``y + 2*x >= 4``. In the nonnegative quadrant,
+    this inequality describes a area above a triangle with vertices at
+    (0, 4), (0, 0) and (2, 0). The minimum of ``f`` occurs at (2, 0).
+    Define A, B, C, D for the standard minimization:
+
+    >>> A, B, C, D = [Matrix(i) for i in [[[2, 1]], [4], [[1, 1]], [-1]]]
+
+    Since `_simplex` will do a minimization for constraints given as
+    ``A*x <= B``, the signs of each are negated (``-Ax >= -B``):
+
+    >>> _simplex(-A, -B, C, D)
+    (3, [2, 0], [1/2])
+
+    The dual of minimizing ``f`` is maximizing ``F = c*y - d`` for
+    ``a*y <= b`` where ``a``, ``b``, ``c``, ``d`` are derived from the
+    transpose of the matrix representation of the standard minimization:
+
+    >>> tr = lambda a, b, c, d: [i.T for i in (a, c, b, d)]
+    >>> a, b, c, d = tr(A, B, C, D)
+
+    This time ``a*x <= b`` is the expected inequality for the `_simplex`
+    method, but to maximize ``F``, the sign of ``c`` and ``d`` must be
+    inverted (so that minimizing the negative will give the negative of
+    the maximum of ``F``):
+
+    >>> _simplex(a, b, -c, -d)
+    (-3, [1/2], [2, 0])
+
+    The negative of ``F`` and the min of ``f`` are the same. The dual
+    point `[1/2]` is the value of ``y`` that minimized ``F = c*y - d``
+    under constraints a*x <= b``:
+
+    >>> y = Matrix(['y'])
+    >>> (c*y - d)[0]
+    4*y + 1
+    >>> [i <= j for i, j in zip(a*y,b)]
+    [2*y <= 1, y <= 1]
+
+    In this 1-dimensional dual system, the more restrictive contraint is
+    the first which limits ``y`` between 0 and 1/2 and the maximum of
+    ``F`` is attained at the nonzero value, hence is ``4*(1/2) + 1 = 3``.
+
+    In this case the values for ``x`` and ``y`` were the same when the
+    dual representation was solved. This is not always the case (though
+    the value of the function will be the same).
+
+    >>> l = [[1, 1], [-1, 1], [0, 1], [-1, 0]], [5, 1, 2, -1], [[1, 1]], [-1]
+    >>> A, B, C, D = [Matrix(i) for i in l]
+    >>> _simplex(A, B, -C, -D)
+    (-6, [3, 2], [1, 0, 0, 0])
+    >>> _simplex(A, B, -C, -D, dual=True)  # [5, 0] != [3, 2]
+    (-6, [1, 0, 0, 0], [5, 0])
+
+    In both cases the function has the same value:
+
+    >>> Matrix(C)*Matrix([3, 2]) == Matrix(C)*Matrix([5, 0])
+    True
+
+    See Also
+    ========
+    _lp - poses min/max problem in form compatible with _simplex
+    lpmin - minimization which calls _lp
+    lpmax - maximimzation which calls _lp
+
+    References
+    ==========
+
+    .. [1] Thomas S. Ferguson, LINEAR PROGRAMMING: A Concise Introduction
+           web.tecnico.ulisboa.pt/mcasquilho/acad/or/ftp/FergusonUCLA_lp_args.pdf
+    """
+
+    A, B, C, D = [Matrix(i) for i in (A, B, C, D or [0])]
+    if dual:
+        _o, d, p = _simplex(-A.T, C.T, B.T, -D)
+        return -_o, d, p
+
+    if A and B:
+        M = Matrix([[A, B], [C, D]])
+    else:
+        assert not A and not B  # no constraints
```

Comment:
I think this assert is guaranteed by the immediately preceding if (no big deal)

Reply:
Oh, no it isn't. How do we know this is satisfied?

Reply:
It now raises a ValueError